### DISCUSSION TITLE
pigpio needs an update

### DISCUSSION THREAD

**Discussion by RonaldTHerman** [Pi4J/pi4j#328] [General]
I'm trying to read from and I2C device that requires a 16 bit register address. I saw something that suggests pi4j has been changed to allow such but apparently pipio doesn't..Here is the error I get

> Exception in thread "main" java.lang.IllegalArgumentException: PIGPIO ERROR: INVALID I2C REGISTER [4400]; Valid range: 0-255
        at com.pi4j.library.pigpio.impl.PiGpioBase.validateI2cRegister(PiGpioBase.java:290)

The base I2C address is 0x11. Within that module is a register at 0x30 giving a full address of 0x00110030

any help appreciated.

**Comment by taartspi** [comment]
That newer interface is                                                                                                                                                                                . default int readRegister(byte[] register, byte[] buffer) {
        return this.readRegister(register, buffer, buffer.length);
    }
public int readRegister(byte[] register, byte[] buffer, int offset, int length)                                                                                                                
                                                                                                                                                                                                                               
 Only available when using the linuxfs provider.  "linuxfs-i2c"                                                                                                                                     Usage: https://github.com/taartspi/pi4j-example-devices/blob/master/src/main/java/com/pi4j/devices/at24c512/AT24C512.java

**Comment by RonaldTHerman** [comment]
Ok I found the jar download.. need to figure out how to get it to work with the I2c code I have. ---------------------------------Ronald T. ***@***.*** NHOn Feb 21, 2024, at 1:30 PM, Tom Aarts ***@***.***> wrote:﻿
That newer interface is                                                                                                                                                                                . default int readRegister(byte[] register, byte[] buffer) {
return this.readRegister(register, buffer, buffer.length);
}
public int readRegister(byte[] register, byte[] buffer, int offset, int length)
Only available when using the linuxfs provider.  "linuxfs-i2c"                                                                                                                                     Usage: https://github.com/taartspi/pi4j-example-devices/blob/master/src/main/java/com/pi4j/devices/at24c512/AT24C512.java
